Tests and files for creators

urbantumbleweed opened the pull request:
#### What's this PR do?

This adds a new directory for action creators.  Code comprising the creators logic was moved out of their corresponding actions files into their own files.  Action creators for errors were created where needed.  In the creators directory is a '**test**' directory where all the creator tests are located.
#### What are the important parts of the code?

For an overview of the code start with the index.js file in creators directory.
#### How should this be tested by the reviewer?

`git pull origin creators && npm install && npm run test`
#### Is any other information necessary to understand this?

creators are used within action files.  The action determines whether there is only a success case or if there is also an error case.

urbantumbleweed committed:
moved action creators out of actions and into creators directory.  Each action has a success case.  Those actions that can throw errors also have a corresponding error creator.

urbantumbleweed committed:
added tests for error creators and corrected a syntax error in creators/index

urbantumbleweed committed:
removed a lone single quote and added an extra line in creator file

urbantumbleweed committed:
created tests for success action creators

urbantumbleweed committed:
added tests for added NEW_FILE and RESET_FILE cases

urbantumbleweed committed:
added test for case that FileReader.readAsText cannot process the input

urbantumbleweed committed:
refactored code: changed 'creaction' to 'action', made consistent the use of string interpolation in 'it' blocks, removed beforeEach and afterEach in favor of const where valued would not change, removed custom errors for errors tests

urbantumbleweed commented:
I refactored the code to incorporate the changes you wanted.  Most notably is the removal of many beforeEach and afterEach blocks because the tests don't change anything.  i still need to write tests to ensure that that additional properties on the action object fail an error.

urbantumbleweed committed:
added tests to ensure that action and action.payload properties are accounted for

urbantumbleweed commented:
I added the tests we talked about that would fail in the event we add a property to the action or payload objects and don't have tests for them.  I used `lodsash.isEqual()` so the properties listed in the test will need to match the order that occurs in the function.  I have alternative approach to this if we must allow the properties to exist in any order but it would be more expensive.

phated commented on src/creators/__tests__/change-directory-error.js:
```diff
@@ -0,0 +1,62 @@
+
+
+describe(`changeDirectoryError creator`, function(){
+  const actionProperties = [
```
these are needed to be tested.  isFSA already tests this

phated commented on src/creators/__tests__/change-directory-error.js:
```diff
@@ -0,0 +1,62 @@
+    done();
+  });
+
+  it(`returns an action object that only has ${actionProperties.length} known ${actionProperties.length === 1 ? 'property' : 'properties'}`, function(done){
```
not needed

phated commented on src/creators/__tests__/change-directory-error.js:
```diff
@@ -0,0 +1,62 @@
+    done();
+  });
+
+  it(`returns an action.payload object that only has ${payloadProperties.length} known ${payloadProperties.length === 1 ? 'property' : 'properties'}`, function(done){
```
don't like this variable substitution, `action.payload has only expected properties`

phated commented on src/creators/__tests__/change-directory-error.js:
```diff
@@ -0,0 +1,62 @@
+  const dirpath = 'change/directory/error/';
+  const err = new Error('Change dir error');
+  const action = changeDirectoryError(dirpath, err);
+  const actionKeys = keys(action);
```
not needed

phated commented on src/creators/__tests__/change-directory-error.js:
```diff
@@ -0,0 +1,62 @@
+  });
+
+  it(`returns an action.payload object that only has ${payloadProperties.length} known ${payloadProperties.length === 1 ? 'property' : 'properties'}`, function(done){
+    expect(isEqual(payloadKeys, payloadProperties)).toEqual(true);
```
this should probably just be 3 expects:

``` js
expect(keys(action.payload)).toInclude('notification');
expect(keys(action.payload)).toInclude('status');
expect(keys(action.payload)).toInclude('error');
```

That can be DRYed up a bit, but it makes it more clear

phated replied:
Scratch that, it doesn't test if there are other keys.

Maybe you could do _.omit() and then check length?

phated replied:
Equal fails on an arrays with keys not in the same order

urbantumbleweed committed:
removed tests on action properties, refactored tests on payload properties, changed descriptions to remove variable interpolation

urbantumbleweed commented:
Code has been refactored and cleaned-up.  I removed the action check, used `_.pick()` for the payload property check, and moved the expectedProps array into the appropriated function.
